### ISSUE TITLE
CMake config header

### ISSUE THREAD

**Issue by jbrezmorf** [flow123d/flow123d#176]
Use generated config.h to pass all necessary data from CMake to compiler.
Namely,
-# information currently in rev_num.h
-# compiler flags
-# various definitions

TODO:
- [x] Implement a CMake macro that can replace add_definition, adding definitions into a list

- [x] Implement a macro that creates config.h from the list

- [x] Test if the file (config.h) is rebuild every time. Since we need to call git to extract actual branch info. Must extract these info every time the CMakeList.txt is processed.

- [x] Define release version in CMake and use it consistently in: install packages names,
version info reported by Flow123d, user manual.

- [x] Report correct compiler flags in `flow123d --version`

**Comment by x3mSpeedy** [comment]
Put together some solution. Changes can be see [here] (https://github.com/x3mSpeedy/flow123d/compare/x3mSpeedy:JHy-coverage-fix...JHy-config_header?expand=0)

Every flag in ```add_definitions``` starting with ```-D``` is stored in directory variable so CMakeLists.txt doesn't have to be changed atleast not much.

I created macro which can replace ```add_definitions```. This macro called ```store_definitions``` is able to capture ALL variables not only those starting with ```-D```
There is also macro which stores other variables (such as flags)

Flow123d now reports correct compiler flags:
```
$ flow123d --version

This is Flow123d, version 1.8.0_master revision: release_1.8.1-262-g31acb7a
Branch: 1.8.0_master
Build: Feb 10 2015, 15:57:43 flags: -g -O0 -Wall -Wno-unused-local-typedefs -std=c++11
Fetch URL:  https://github.com/flow123d/flow123d.git
```


**config.h** looks like this now:

```c++
// This is autogenerated file. Do not modify this file!
// File contains compiler flags and other definitions.

#define _COMPILER_FLAGS_ "-g -O0 -Wall -Wno-unused-local-typedefs -std=c++11"
#define _GIT_REVISION_ "release_1.8.1-262-g31acb7a"
#define _GIT_BRANCH_ "1.8.0_master"
#define _GIT_URL_ " https://github.com/flow123d/flow123d.git"
#define _VERSION_NAME_ "1.8.0_master"
#define CMAKE_CXX_FLAGS "-g -O0 -Wall -Wno-unused-local-typedefs -std=c++11"
#define CMAKE_C_FLAGS "-g -O0"
#define CMAKE_EXE_LINKER_FLAGS "-g  -Wl,-rpath,$ORIGIN/../lib"
```

**Comment by jbrezmorf** [comment]
I prefer to use single dedicated macro "flow_define" to add configuration definitions and pass these only through the configuration file in order to cleanup compiler parameters.
In fact there are at least two kind of definitions:
1) flags indicating e.g. presence of some library:
	flow_define("HAVE_PETSC")
results in:
	#define FLOW123D_HAVE_PETSC=1
2) string defines
	flow_define("GIT_URL" "http://...")
results in:
	#define FLOW123D_GIT_URL="http://..."
This can be done with single macro with second parameter optional.

**Comment by x3mSpeedy** [comment]
I changed macro (0e7f32152a9c553912bf65aeef0cddd9634cf6b1) a bit (now called ```flow_define```) to support behaviour specfied above. If called in CMake script:
```
flow_define (HAVE_PETSC)
flow_define (GIT_URL Foo)
```
will produce lines
```c++
#define FLOW123D_HAVE_PETSC 1
#define FLOW123D_GIT_URL "Foo"
```

I use python script for expanding definition in list. If current demands for this macro won't change, it can be done also in CMake but if we are to add other types of definitions like **number defines** or other, distinguishing them in CMake script wont be easy task.

I notices that ```rev_num.h``` procedure is being done by creating custom target and adding dependency to flow213d. Should creating of ```config.h``` be done in same fashion?

**Comment by jbrezmorf** [comment]
On 02/11/2015 11:41 AM, Jan Hybš wrote:
> I changed macro (0e7f321
> <https://github.com/flow123d/flow123d/commit/0e7f32152a9c553912bf65aeef0cddd9634cf6b1>)
> a bit (now called |flow_define|) to support behaviour specfied above. If
> called in CMake script:
> 
> |flow_define (HAVE_PETSC)
> flow_define (GIT_URL Foo)
> |
> 
> will produce lines
> 
> #define FLOW123D_HAVE_PETSC 1
> #define FLOW123D_GIT_URL "Foo"
> 
> I use python script for expanding definition in list. If current demands
> for this macro won't change, it can be done also in CMake but if we are
> to add other types of definitions like *number defines* or other,
> distinguishing them in CMake script wont be easy task.

The expansion can be done on few lines using CMAKE, no need to create
separate script with own parsing. This also breaks readability.

For possible numeric or constant defines we may introduce
flow_define_constant(...) in future.

> 
> I notices that |rev_num.h| procedure is being done by creating custom
> target and adding dependency to flow213d. Should creating of |config.h|
> be done in same fashion?

Absolutely necessary, otherwise flow123d will not rebuild if the
config.h is changed.

> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flow123d/flow123d/issues/176#issuecomment-73862060>.
> 

-- 
------------------------
Mgr. Jan Brezina, Ph. D.
Technical University in Liberec, New technologies institute
http://www.nti.tul.cz/cz/WikiUser:Jan.Brezina

**Comment by x3mSpeedy** [comment]
I removed python script and did simple solution in CMake.
I also added custom target for creating config.h

During execution of main ```CMakeLists.txt``` is always created tmp config file containing all definitions. Upon starting ```custom_target``` are added git details. Reason why is nor everything created from the scratch in ```custom_target``` phase is because handing over all declared definitions thru parameter is given its content (spaces, semicolons, equal sign, quotation marks) could be quite complex.

```config.h``` is generated only if there's some change in it (```copy_if_different```).

Testing results seems ok (measurements are approximate), running make all:
 - **without** any changes takes  ```7 sec```
 - **with** changed config.h takes ```20 sec``` 
 - **with** changed num_rev.h takes also ```20 sec```
 - **with** changed any src file takes also ```20 sec```

I had to rename constant in main.cc since flow_define automatically adds prefix FLOW123D 

Take a look at all [changes] (https://github.com/x3mSpeedy/flow123d/compare/x3mSpeedy:JHy-coverage-fix...JHy-config_header?expand=0)
